In [1]:
import pandas as pd
import seaborn as sns
import matplotlib as plt
import numpy as np

### Importing the Data

The data is comma seperated, and the raw data being read from absolulute directory /data/raw/

In [63]:
df = pd.read_csv("../data/raw/autoscout24-germany-dataset.csv")

display(df)

mileage        make   model               fuel    gear       offerType  \
0       235000         BMW     316             Diesel  Manual            Used   
1        92800  Volkswagen    Golf           Gasoline  Manual            Used   
2       149300        SEAT    Exeo           Gasoline  Manual            Used   
3        96200     Renault  Megane           Gasoline  Manual            Used   
4       156000     Peugeot     308           Gasoline  Manual            Used   
...        ...         ...     ...                ...     ...             ...   
46400       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46401       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46402       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46403       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46404       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   

       price     hp  year  
0       6800  116.0  2011  
1       6877  122.0  2011  
2       6900  160.0  2011  
3       6950  110.0  2011  
4       6950  156.0  2011  
...      ...    ...   ...  
46400  12990   71.0  2021  
46401  12990   71.0  2021  
46402  12990   71.0  2021  
46403  12990   71.0  2021  
46404  12990   71.0  2021  

[46405 rows x 9 columns]

### Cleaning Data

The data is 99.3% complete, but unfortunately there are 334 rows (out of 46405 rows) that contain incomplete information about the vehicle.
- Mileage, make, fuel, offerType, price, and year contain complete information.
- 143 rows contain missing vehicle model, for these cases I am going to replace the blank space with "Unknown" since the data associated with the vehicle is still valuable even if the model is unkown.
- 182 rows contain missing transmission information, these will be dropped since this column will be converted to a numeric later.
- 29 vehicles are missing their horsepower, I am going to delete these few rows since they are inconsequencial compared to the size of the data set and provide little use with such an impactful metric missing
- There are 23 vehicles with brand "Others", these vehicles provide no insight into make/model or any other correlations that can be drawn thus they are dropped from the dataframe.
- In between each of the row drop operations, the row index's are re-indexed from 0 for the next row drop operation.
- I find "gear" a little odd, I prefer to use "transmission" so I am going to rename the column. Additionally I am going to assign numeric scores to the transmission category. Manual is being converted to 5, semi-automatic to 7.5 and automatic to 10; this is needed for the next step.

#### My Scoring Metric

I am interested in purchasing a new vehicle and I find certain aspects of a vehicle more important than others. I am going to implement a multiple account analysis/decision matrix approach in order to rank each vehicle against a weighting. Each vehicle's attribute will be first normalized against the rest of the data set then the specific attribute will be weighted according to the pre-described weightings. Year is the most important factor when vehicle shopping thus it is assigned a weighting score of 4. Next is price, naturally one shops according to a budget thus it is weighted as 2. Next is mileage, a vehicle with low mileage is generally more reliable thus is scored 2. Horsepower is an attribute some vehicle shop for, but for this scoring metric it ranks low compared to other factors, earning it a 1 weighting factor. Finally is transmission, giving it a weight of 1.

In [64]:
def f(row):
    if( row["gear"] == "Manual"):
        val = 5
    elif row["gear"] == "Semi-automatic":
        val = 7.5
    else: 
        val = 10
    return val

In [85]:
df = pd.read_csv("../data/raw/autoscout24-germany-dataset.csv")

df["model"] = df["model"].fillna("Unknown")

df = df.drop(df[df["gear"].isna()].index)

df = df.reset_index()
df = df.drop("index", axis = "columns")

df = df.drop(df[df["hp"].isna()].index)

df = df.reset_index()
df = df.drop("index", axis = "columns")

df = df.drop(df[df["make"] == "Others"].index)

df = df.reset_index()
df = df.drop("index", axis = "columns")

df["transmission"] = df.apply(f, axis = 1) 
df = df.drop("gear", axis = "columns")

df = df[["mileage", "make", "model", "fuel", "transmission", "offerType", "price", "hp", "year"]]

temp = df

temp["inverse_mileage"] = 1/df["mileage"]
temp["inverse_transmission"] = 1/df["transmission"]
temp["inverse_price"] = 1/df["price"]
temp["inverse_hp"] = 1/df["hp"]
temp["inverse_year"] = 1/df["year"]
temp["score"] = 4 * temp["year"] * 10 * temp["inverse_year"].min() + 2 * temp["inverse_price"] * 10 * temp["price"].min() + 2 * temp["inverse_mileage"] * 10 * temp["mileage"].min() + 1 * temp["hp"] * 10 * temp["inverse_hp"].min() + 1 * temp["transmission"] * 10 * temp["inverse_transmission"].min()
print(temp.keys())
display(temp.sort_values(by = "score", ascending = False))

mileage        make   model               fuel transmission  \
0       235000         BMW     316             Diesel       Manual   
1        92800  Volkswagen    Golf           Gasoline       Manual   
2       149300        SEAT    Exeo           Gasoline       Manual   
3        96200     Renault  Megane           Gasoline       Manual   
4       156000     Peugeot     308           Gasoline       Manual   
...        ...         ...     ...                ...          ...   
46175       99        Fiat     500  Electric/Gasoline       Manual   
46176       99        Fiat     500  Electric/Gasoline       Manual   
46177       99        Fiat     500  Electric/Gasoline       Manual   
46178       99        Fiat     500  Electric/Gasoline       Manual   
46179       99        Fiat     500  Electric/Gasoline       Manual   

            offerType  price     hp  year  
0                Used   6800  116.0  2011  
1                Used   6877  122.0  2011  
2                Used   6900  160.0  2011  
3                Used   6950  110.0  2011  
4                Used   6950  156.0  2011  
...               ...    ...    ...   ...  
46175  Pre-registered  12990   71.0  2021  
46176  Pre-registered  12990   71.0  2021  
46177  Pre-registered  12990   71.0  2021  
46178  Pre-registered  12990   71.0  2021  
46179  Pre-registered  12990   71.0  2021  

[46180 rows x 9 columns]